feat: Speedup unlink operation for big directories

dmga44 opened the pull request:
This PR mainly targets the unlink operation performance on big directories. The key change was to replace the O(n) [n = number of entries in the directory] previous implementation of getChildName to other one that behaves O(1) for non-hardlinked files.

There are some other side changes:
- modify the big session metadata benchmark to show results per every chunk of files (not yet configurable).
- speedup the dumping of metadata due to now creating the buffer to store nodes and edges every time. These buffers were not small.
- increase the size of the chunk hash table used in master to make their linked lists shorter. This change targets speeding up the traversing of those linked lists.

dmga44 committed:
feat: Make MD benchmark to show time per slice

This change makes possible to analyze the performance of the metadata
operations considering the different sizes of the folder it is
currently worked on. For now the results every 10000 is being shown to
make easier the calculations on operations per second.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
feat: Reuse the buffers when storing metadata

The goal of this change is to speed up the storing of metadata by
saving the allocation of the buffers used to write the nodes and edges
metadata before writing those into the metadata file. Also some more
explicit naming of the length constants were used to make code more
readable.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
feat: Reduce unlink overhead

Currently unlink syscall was calling several times the function
FSNodeDirectory::getChildName. That function was taking time
proportional to the number of entries inside the directory, which can
make the unlink operation in big directories to be very expensive.

This commit makes those operations taking time proportional to the
number of parents of the node to get the name of. The idea is to store,
beside the inode of the parents of a node, also a pointer to the handle
of the name of the node inside the parent entries structure.

As a side change, the case insensitive directories were optimized to be
O(logn) in the find queries instead of previous O(n).

Signed-off-by: Dave <dave@leil.io>

antuan96314 reviewed (dismissed):
LGTM

lgsilva3087 commented on src/master/filesystem_node_types.h:
```diff
@@ -203,17 +206,19 @@ struct FSNodeDevice : public FSNode {
-	HandleCompare> EntriesContainer;
-#endif
+
+	typedef __gnu_pbds::tree<std::pair<hstorage::Handle, FSNode *>,
+	                         __gnu_pbds::null_type, HandleCompare,
+	                         __gnu_pbds::rb_tree_tag,
+	                         __gnu_pbds::tree_order_statistics_node_update>
+	    EntriesContainer;
```
How is the portability of this extensions? Are they available only for GCC or also in Clang (for instance)?

dmga44 replied:
This change was moved to PR #272.

lgsilva3087 reviewed (commented):
Great job. Please see my question.

dmga44 committed:
feat: Enlarge chunk hash table

The enlarge of this table targets shorter linked per each hash
position, which possibly impact positively in the speed of metadata
loading.

Signed-off-by: Dave <dave@leil.io>

rolysr reviewed (approved):
LGTM

ralcolea commented on src/master/filesystem_node.cc:
```diff
@@ -71,6 +71,9 @@ FSNode *FSNode::create(uint8_t type) {
 }
 
 void FSNode::destroy(FSNode *node) {
+	for (auto const &[_ /*: parentId*/, handlePtr] : node->parent) {
```
I like the idea of using the placeholder (`_`) and remove `/*: parentId/*` to be consistent with other parts of the code.

dmga44 replied:
Applied.

ralcolea commented on src/master/filesystem_node.cc:
```diff
@@ -509,29 +519,55 @@ void fsnodes_remove_edge(uint32_t ts, FSNodeDirectory *parent, const HString &na
-	if (it != node->parent.end()) {
-		node->parent.erase(it);
-	}
+	if (!parent->case_insensitive) {
```
Consider refactoring this code to avoid duplicated scenarios with very similar logic:
```cpp
HString currentName = name;
if (parent->case_insensitive) {
	currentName = HString::hstringToLowerCase(name);
}

auto it = std::find_if(
    node->parent.begin(), node->parent.end(),
    [parent,
     currentName](const std::pair<uint32_t, const hstorage::Handle *> &p) {
	    return p.first == parent->id && p.second->get() == currentName;
    });

if (it != node->parent.end()) { 
      node->parent.erase(it);
}
```

dmga44 replied:
Please notice there is another change which you don't consider when case insensitive:
```return p.first == parent->id && HString::hstringToLowerCase(p.second->get()) == lowerCaseName;```
the ```hstringToLowerCase``` function is not applied in your proposal, and it is needed.

I think it can be applied, more or less.

ralcolea commented on src/master/filesystem_node_types.h:
```diff
@@ -123,8 +123,10 @@ struct FSNode {
-	compact_vector<uint32_t, uint32_t> parent; /*!< Parent nodes ids. To reduce memory usage ids
-	                                                are stored instead of pointers to FSNode. */
+	compact_vector<std::pair<uint32_t, const hstorage::Handle *>, uint32_t>
+	    parent; /*!< Parent nodes ids + name of the of entries in those parents.
```
Please, fix some typos in the description of `parent` attribute.
I think the description should reference `handle` instead of `names`, even if in practice, `Handle` behaves like a name, but I think `Handle` it's more general and contains more information.

dmga44 replied:
Applied.

ralcolea commented on src/master/filesystem_node_types.h:
```diff
@@ -203,17 +205,17 @@ struct FSNodeDevice : public FSNode {
  */
 struct FSNodeDirectory : public FSNode {
-#if defined(SAUNAFS_HAVE_64BIT_JUDY) && !defined(DISABLE_JUDY_FOR_ENTRIESCONTAINER)
-	typedef judy_map<hstorage::Handle, FSNode *> EntriesContainer;
```
Why is it removed `judy_map` from `FSNodeDirectory`?
What consequences can have this change in the performance of operations like insertion, deletion and lookup?
Why is it removed only from `FSNodeDirectory` and not from other data types?

dmga44 replied:
Let's go one by one:
- the ```judy_map``` was removed mainly because the type of the keys in the map changed from ```Handle``` to ```Handle *```, and having it would imply coding more things. Also, the ```judy_map``` was not used in the last releases, and was substituted by the ```flat_map``` which is used here.
- the motivation of the substitution of ```judy_map``` by ```flat_map``` was increasing the performace of insertion, deletion and lookup. Also, the PR #272 replaces this data structure again.
- in the other places where the ```judy_map``` is used, I have not seen any reason to remove it or not use it. Substituting it in some places, like in the ```SparseBitset``` structure may have terrible consecuences.

ralcolea replied:
I remembered it was the other way around, I meant, judy_map was faster than flat_map, that's why my question. Your third idea is closer to what I remembered :)

ralcolea commented on src/master/filesystem_periodic.cc:
```diff
@@ -417,8 +418,12 @@ void fs_process_file_test() {
 					if (!node ||
-					    std::find(node->parent.begin(), node->parent.end(),
-					              f->id) == node->parent.end()) {
+					    std::find_if(
```
Consider storing the result of `std::find_if()` function in a different variable.
The mixture of `std::find_if` with the lambda and the comparison with `node->parent.end()`
is less readable and more difficult to read.

dmga44 replied:
Notice the first part of the or operator, ```!node``` ensures that when the first part is not met, and the second is ran, the node always exist. That's why it was not looking that bad before, I've uploaded a change that maybe looks better, comments added :)

ralcolea reviewed (commented):
Very nice changes @dmga44 :+1: :fire: :muscle: 
Please, see my suggestions and questions.

ralcolea reviewed (approved):
Great job @dmga44 :+1: :fire: :muscle:
